packages/back-end/src/routers/dashboards/dashboards.controller.ts

```diff
@@ -0,0 +1,304 @@
+  res: ResponseWithStatusAndError<SingleDashboardResponse>
+) {
+  const context = getContextFromReq(req);
+  if (!context.hasPremiumFeature("dashboards")) {
+    throw new Error("Must have a commercial License Key to manage Dashboards");
+  }
+
+  const { id } = req.params;
+  const updates = { ...req.body };
+  const dashboard = await context.models.dashboards.getById(id);
+  if (!dashboard) throw new Error("Cannot find dashboard");
+  const experiment = await getExperimentById(context, dashboard.experimentId);
+  if (!experiment) throw new Error("Cannot find connected experiment");
+
+  if (updates.blocks) {
+    // Duplicate permissions checks to prevent persisting the child blocks if the user doesn't have permission
+    const isOwner = context.userId === dashboard.userId || !dashboard.userId;
+    const isAdmin = context.permissions.canSuperDeleteReport();
+    const canEdit =
+      isOwner ||
+      isAdmin ||
+      (dashboard.editLevel === "organization" &&
+        context.permissions.canUpdateReport(experiment));
+    const canManage = isOwner || isAdmin;
+
+    if (!canEdit) context.permissions.throwPermissionError();
+    if (
+      ("title" in updates ||
+        "editLevel" in updates ||
+        "enableAutoUpdates" in updates) &&
+      !canManage
+    ) {
+      context.permissions.throwPermissionError();
+    }
+
+    const createdBlocks = await Promise.all(
+      updates.blocks.map((blockData) =>
+        isPersistedDashboardBlock(blockData)
+          ? blockData
+          : createDashboardBlock(context.org.id, blockData)
+      )
+    );
+    updates.blocks = createdBlocks;
+  }
+
+  const updatedDashboard = await context.models.dashboards.updateById(
+    id,
+    updates as Partial<DashboardInstanceInterface>
+  );
+
+  res.status(200).json({
+    status: 200,
+    dashboard: updatedDashboard,
+  });
+}
+
+export async function deleteDashboard(
+  req: AuthRequest<never, { id: string }, never>,
+  res: ResponseWithStatusAndError
+) {
+  const context = getContextFromReq(req);
+  const { id } = req.params;
+  await context.models.dashboards.deleteById(id);
+  return res.status(200).json({ status: 200 });
+}
+
+export async function refreshDashboardData(
+  req: AuthRequest<never, { id: string }, never>,
+  res: ResponseWithStatusAndError
+) {
+  const context = getContextFromReq(req);
+  const { id } = req.params;
+  const dashboard = await context.models.dashboards.getById(id);
+  if (!dashboard) throw new Error("Cannot find dashboard");
+  const experiment = await getExperimentById(context, dashboard.experimentId);
+  if (!experiment)
+    throw new Error("Cannot update dashboard without an attached experiment");
+  const datasource = await getDataSourceById(context, experiment.datasource);
+  if (!datasource) throw new Error("Failed to find connected datasource");
+
+  // This is doing an expensive analytics SQL query, so may take a long time
+  // Set timeout to 30 minutes
+  req.setTimeout(SNAPSHOT_TIMEOUT);
```

Comment:
I think if we are adopting the pattern of `create a query, return the id, provide and endpoint to get the latest status` then we can remove this increased timeout.

Otherwise we do need to wait on line `executeAndSaveQuery` otherwise this is also not needed and we should remove it.